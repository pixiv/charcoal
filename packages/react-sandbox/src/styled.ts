import styled from 'styled-components'
import createTheme from '@charcoal/styled'
export const theme = createTheme(styled)
