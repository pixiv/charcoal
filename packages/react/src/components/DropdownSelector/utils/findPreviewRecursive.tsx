import React, { ReactNode } from 'react'

/**
 * DropdownSelectorの選択中の要素をレンダリングするため、
 * 選択中のMenuItemを再帰的に探索しReactNodeを返す。
 *
 * @param children
 * @param value
 * @param values
 * @returns
 */
export function findPreviewRecursive(
  children: ReactNode,
  value: string
): ReactNode | undefined {
  const childArray = React.Children.toArray(children)
  for (let i = 0; i < childArray.length; i++) {
    const child = childArray[i]
    if (React.isValidElement(child)) {
      if ('value' in child.props) {
        const childValue = (child.props as { value: string }).value
        if (childValue === value && 'children' in child.props) {
          const children = (child.props as { children: ReactNode }).children
          return children
        }
      }
      if ('children' in child.props) {
        const children = findPreviewRecursive(
          (child.props as { children: ReactNode }).children,
          value
        )
        if (children !== undefined) {
          return children
        }
      }
    }
  }
}
