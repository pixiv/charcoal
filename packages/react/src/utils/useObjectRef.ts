import { useCallback, useRef } from 'react'

/**
 * Custom implementation of useObjectRef to replace @react-aria/utils
 * Converts any ref to a mutable ref object
 */
export function useObjectRef<T>(
  forwardedRef?: React.Ref<T>
): React.MutableRefObject<T | null> {
  const objRef = useRef<T | null>(null)

  const ref = useCallback(
    (value: T | null) => {
      objRef.current = value

      if (typeof forwardedRef === 'function') {
        forwardedRef(value)
      } else if (forwardedRef && typeof forwardedRef === 'object') {
        ;(forwardedRef as React.MutableRefObject<T | null>).current = value
      }
    },
    [forwardedRef]
  )

  // Return an object that can be used as both a callback ref and a mutable ref
  return {
    get current() {
      return objRef.current
    },
    set current(value: T | null) {
      ref(value)
    },
  }
}
