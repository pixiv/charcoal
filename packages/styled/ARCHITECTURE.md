# `@charcoal-ui/styled` の仕組み

以下、`@charcoal-ui/styled` を省略して styled と称することがある。

```
theme(( o ) => [ (o.bg.background1.hover.press) ] )
  │   │ │        └───────────────────────────────── Internal 型
  │   │ └────────────────────────────────────────── builder（ "o" ）
  │   └──────────────────────────────────────────── spec 関数
  └──────────────────────────────────────────────── theme 関数
```

## 用語集

※ 用語は 2023-03-09 時点のものです。命名に改善の余地はあると思う。

### `theme()` 関数

`@charcoal-ui/styled` における DSL を開始する宣言。

この中で囲まれた範囲では、charcoal のガイドラインに沿ったスタイルしか（そのメソッドチェーンに従って）書くことができない。

返り値は `(props: P) => CSSObject` である。theme 関数の型が、ではなく返り値の型がこれである。

つまり、`styled-components` のテンプレートに書くことができる関数を組み立てる高階関数が `theme()` だと言える。

### `spec()` 関数

`@charcoal-ui/styled` において、ユーザーがスタイルを定義する関数。

基本は `o` を受け取って、`Internal`、`Internal[]` を返す。
が、条件によって当たったり当たらなかったりするスタイル定義もあるため、`null` `false` `undefined` も返り値として許容される。

すなわち、spec の返り値は `Internal | null | false | undefined | Array<Internal | null | false | undefined>` である。

### `builder`

spec の引数として現れる、 `o` と名指されるオブジェクトのこと。

`o` は特に何かの略ではなく、箇条書き感のある記号としてつけられている。

中にメソッドチェーンの可能なオブジェクトが、CSS のプロパティ分野ごとに生えている。これの定義方法は後述する。

### `Internal` 型

`o.bg.background1` のような宣言の返り値のこと。`theme()` 関数がこれを解釈し、最終的に `CSSObject` へと変換する。

ユーザーからは `background1` や `hover` などのメソッドしか生えていないように見えるが、実際には `[internalSym]` という `Symbol` 型のプロパティが生えている。

このシンボルはライブラリ外に export されないため、ユーザーはこのプロパティを叩くことはできないが、`theme()` 関数はこれにアクセスできるという事実上パッケージプライベートな挙動をする。

### `Context` 型

複数のスタイル定義の間（というより 1 つの `theme()` 関数内）で共有される状態のこと。たとえば

- `o.bg.background1.hover` を書くと自動で `transition: background-color 0.2s` が挿入される（ホバー時に transition するため）
- `o.typography(14).preserveHalfLeading` を書くと自動で `::before` `::after` が挿入される

といった挙動を実現するのに使われる。

`context` は各 `Internal` 型の `internalSym` の中にプロパティとして生える。

## メソッドチェーンの仕組み

`Internal` 型は、メソッドチェーンの続きにあたるメソッドと、現時点で確定したスタイル定義を格納した `internalSym` を持つ。

たとえば、`o.bg.background1` の中身はおおよそ次のようになる。

```ts
return {
  // このあと叩いていいメソッド
  get hover() {
    ...
  },
  get press() {
    ...
  },

  // 内部状態（ ライブラリ内からしかアクセスできない ）
  [internalSym]: {
    // 複数のスタイル定義で共有する状態
    context: {},

    // CSS を組み立てる関数
    toCSS(context: Context): CSSObject
  }
}
```

`Internal` を定義するのに使われるのが `define{Propertyies,Methods,PropertyChains,MethodChains}` である。

1 つの factory が 1 つのメソッドに対応するとは限らない（1 度に複数のメソッドを定義する factory もある）。内部的には最終的に `Object.defineProperties()` が行われる。

詳しくは `factories/lib.ts` を参照。
