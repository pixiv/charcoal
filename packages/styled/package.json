{
  "name": "@charcoal-ui/styled",
  "version": "2.6.0",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    "require": "./dist/index.cjs.js",
    "default": "./dist/index.esm.js"
  },
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "run-p --print-label 'build:*'",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "typecheck": "tsc --project tsconfig.build.json --pretty --noEmit",
    "clean": "rimraf dist .tsbuildinfo"
  },
  "devDependencies": {
    "@types/react": "^17.0.38",
    "@types/react-test-renderer": "^17.0.2",
    "@types/styled-components": "^5.1.21",
    "@types/warning": "^3.0.0",
    "jest-styled-components": "^7.1.1",
    "npm-run-all": "^4.1.5",
    "react": "^17.0.2",
    "react-test-renderer": "^17.0.2",
    "rimraf": "^3.0.2",
    "styled-components": "^5.3.3",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@charcoal-ui/foundation": "^2.6.0",
    "@charcoal-ui/theme": "^2.6.0",
    "@charcoal-ui/utils": "^2.6.0",
    "warning": "^4.0.3"
  },
  "peerDependencies": {
    "react": ">=16.13.1",
    "styled-components": ">=5.1.1"
  },
  "files": [
    "src",
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/styled"
  },
  "gitHead": "8579b406b316285a35858512030d2143524ae154"
}
