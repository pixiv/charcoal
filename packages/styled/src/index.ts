import { CSSObject, ThemedStyledInterface } from 'styled-components'
import { CharcoalAbstractTheme } from '@charcoal-ui/theme'
import { ArrayOrSingle, isPresent, noThemeProvider, wrapArray } from './util'
import { Internal, toCSSObjects } from './internals'
import createO from './builders/o'
import transition from './builders/transition'
export { default as TokenInjector } from './TokenInjector'
export {
  getThemeSync,
  themeSetter,
  themeSelector,
  prefersColorScheme,
  useTheme,
  useThemeSetter,
  useLocalStorage,
  useMedia,
} from './helper'
export { defineThemeVariables } from './util'
export { removeHalfLeadingCss } from './utils/typographyCss'
export * from './SetThemeScript'

type Blank = null | undefined | false

const nonBlank = <T>(value: T): value is T extends Blank ? never : T =>
  isPresent(value) && (value as unknown) !== false

/**
 * @deprecated
 * 実行時のパフォーマンスが低い問題があります。
 * ref. https://github.com/pixiv/charcoal/pull/377
 *
 * styled-components の theming 方法を推奨します。
 * ref. https://styled-components.com/docs/advanced#theming
 *
 * `theme(o => [...])` の `theme` ユーティリティを構築する
 *
 * @param _styled - DEPRECATED: styled-components の `styled` そのものを渡すとそれを元に型推論ができる。が、型引数を渡す方が型推論が高速になりやすい
 *
 * @example
 *
 * import styled from 'styled-components'
 * const theme = createTheme(styled)
 *
 * @example
 *
 * const theme = createTheme<DefaultTheme>()
 */
export function createTheme<T extends CharcoalAbstractTheme>(
  _styled?: ThemedStyledInterface<T>
) {
  type Builder = ReturnType<typeof createO<T>>

  // ランタイムの `theme(o => [...])` のインターフェースを構築する
  return function theme(
    specFn: (o: Builder) => ArrayOrSingle<Internal | Blank>
  ): ThemeProp<T> {
    // styled-components のテンプレートに埋め込める関数
    return function interpolate({ theme }) {
      if (!isPresent(theme)) {
        // テーマが入っていない場合は復旧不可能なのでエラーにする
        throw noThemeProvider
      }

      const internals = [
        // ユーザーが定義したルール
        ...wrapArray(
          /**
           * こう書いてはいけない
           *
           * ❌
           * ```ts
           * const o = createO(theme)
           * const declaration = spec(o)
           * ```
           *
           * `o` を一時変数に入れてしまうと型 `T` の具象化が行われるので関数内に書く
           */
          specFn(/** o = */ createO(theme))
        ),

        // 必ず挿入される共通のルール
        transition(theme),
      ].filter(nonBlank)

      return toCSSObjects(internals)
    }
  }
}

export type ThemeProp<T> = ({
  theme,
}: {
  theme: T | undefined
}) => CSSObject | CSSObject[]

export { focusVisibleFocusRingCss } from './styles/focusVisibleFocusRingCss'
export { disabledCss } from './styles/disabledCss'
export { assertiveRingCss } from './styles/assertiveRingCss'
export type { CharcoalThemeUtils } from './utils/CharcoalStyledTheme'
export { addThemeUtils } from './utils/addThemeUtils'
