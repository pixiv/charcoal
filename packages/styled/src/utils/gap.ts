import { Spacing } from '@charcoal-ui/theme'
import { css, RuleSet } from 'styled-components'
import { pxIfNum } from './helpers/pxIfNum'

export function gap(v1: keyof Spacing): RuleSet

export function gap(v1: keyof Spacing, v2: keyof Spacing): RuleSet

export function gap(v1: keyof Spacing, v2?: keyof Spacing) {
  return css`
    gap: ${v1}px ${pxIfNum(v2)};
  `
}

export function rowGap(v: keyof Spacing) {
  return css`
    row-gap: ${v}px;
  `
}

export function columnGap(v: keyof Spacing) {
  return css`
    column-gap: ${v}px;
  `
}
