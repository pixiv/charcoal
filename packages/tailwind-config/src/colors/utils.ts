import { GradientMaterial, Material } from '@charcoal/theme'

export const COLOR_PREFIX = '--tailwind-color-'

export function isSingleColor(color: AnyColor): color is Material {
  return typeof color === 'string'
}

type AnyColor = Material | GradientMaterial

export type AnyColorTheme = Record<string, AnyColor>
