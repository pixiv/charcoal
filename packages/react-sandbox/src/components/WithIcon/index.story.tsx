import { boolean } from '@storybook/addon-knobs'
import React from 'react'
import styled, { css } from 'styled-components'
import { theme } from '../../styled'
import WithIcon from '.'
import { applyEffect } from '@charcoal/utils'

export default {
  title: 'Sandbox/WithIcon',
  component: WithIcon,
}

export const Default = () => {
  const fit = boolean('fit', true)
  const bigger = boolean(
    'Icon height bigger than line-height (`fit` is effective)',
    false
  )
  const prefix = boolean('prefix', false)
  const show = boolean('show', true)
  return (
    <Container>
      <WithIcon
        icon={<TestInlineIcon lineHeight={bigger ? 32 : 22} />}
        show={show}
        prefix={prefix}
        fit={fit}
      >
        Menu
      </WithIcon>
    </Container>
  )
}

export const Performance = () => (
  <Container>
    <WithIcon
      icon={<TestInlineIcon lineHeight={22} />}
      show
      fit
      // Hard-coded width (ResizeObserver free and NO measuring cost)
      width={16}
    >
      Menu
    </WithIcon>
  </Container>
)

export const Naive = () => (
  // NO measuring cost
  <Container>
    <WithIcon icon={<TestInlineIcon lineHeight={22} />} show>
      Menu
    </WithIcon>
  </Container>
)

export const Prefix = () => (
  <Container>
    <WithIcon icon={<TestIcon />} show prefix fit>
      Selection
    </WithIcon>
  </Container>
)

function Container(props: { className?: string; children?: React.ReactNode }) {
  return (
    <div
      css={css`
        ${theme((o) => [o.font.text1, o.typography(14).preserveHalfLeading])}
        display: flex;
      `}
    >
      <div
        css={css`
          background-color: ${({ theme }) =>
            applyEffect(theme.color.brand, theme.elementEffect.disabled)};
        `}
        {...props}
      />
    </div>
  )
}

export const Hide = () => (
  <Container>
    <WithIcon icon={<TestIcon />} show={false} prefix>
      Selection
    </WithIcon>
  </Container>
)

export const Collapse = () => (
  <Container>
    <WithIcon icon={<TestIcon />} show="collapse" prefix>
      Selection
    </WithIcon>
  </Container>
)

export const LongText = () => (
  <Container
    css={`
      width: 200px;
    `}
  >
    <WithIcon icon={<TestIcon />}>
      Long Long Long Long Long Long Long Long Long Long Text
    </WithIcon>
  </Container>
)

export const LongTextOverflow = () => (
  <Container
    css={`
      width: 200px;
    `}
  >
    <WithIcon icon={<TestIcon />} fixed>
      Long Long Long Long Long Long Long Long Long Long Text
    </WithIcon>
  </Container>
)

const TestIcon = styled.div`
  display: inline-block;
  width: 16px;
  height: 16px;
  background-color: currentColor;
`

const TestInlineIcon = styled.div<{ lineHeight: number }>`
  display: inline-flex;
  vertical-align: top;
  align-items: center;
  line-height: ${(p) => p.lineHeight}px;
  height: ${(p) => p.lineHeight}px;
  &::before {
    content: '';
    display: inline-block;
    height: 16px;
    width: 16px;
    background-color: currentColor;
  }
`
