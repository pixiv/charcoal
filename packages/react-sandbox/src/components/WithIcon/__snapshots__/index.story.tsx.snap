// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Sandbox/WithIcon Collapse 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  min-width: 0;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: none;
  margin-right: 4px;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c2 > svg {
  display: block;
}

.c3 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css={
        Array [
          "
          background-color: ",
          [Function],
          ";
        ",
        ]
      }
    >
      <div
        className="c0"
      >
        <div
          className="c1"
        >
          <div
            className="c2"
          >
            <div
              className="c3"
            />
          </div>
        </div>
        <div
          className="c4"
        >
          Selection
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Sandbox/WithIcon Default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  min-width: 0;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 0;
  width: 0px;
  visibility: visible;
  margin-left: 4px;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: absolute;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c3 > svg {
  display: block;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 22px;
  height: 22px;
}

.c4::before {
  content: '';
  display: inline-block;
  height: 16px;
  width: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css={
        Array [
          "
          background-color: ",
          [Function],
          ";
        ",
        ]
      }
    >
      <div
        className="c0"
      >
        <div
          className="c1"
        >
          Menu
        </div>
        <div
          className="c2"
          width={0}
        >
          <div
            className="c3"
          >
            <div
              className="c4"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Sandbox/WithIcon Hide 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  min-width: 0;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  visibility: hidden;
  margin-right: 4px;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c2 > svg {
  display: block;
}

.c3 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css={
        Array [
          "
          background-color: ",
          [Function],
          ";
        ",
        ]
      }
    >
      <div
        className="c0"
      >
        <div
          className="c1"
        >
          <div
            className="c2"
          >
            <div
              className="c3"
            />
          </div>
        </div>
        <div
          className="c4"
        >
          Selection
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Sandbox/WithIcon Long Text 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  min-width: 0;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  visibility: visible;
  margin-left: 4px;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c3 > svg {
  display: block;
}

.c4 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css="
      width: 200px;
    "
    >
      <div
        className="c0"
      >
        <div
          className="c1"
        >
          Long Long Long Long Long Long Long Long Long Long Text
        </div>
        <div
          className="c2"
        >
          <div
            className="c3"
          >
            <div
              className="c4"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Sandbox/WithIcon Long Text Overflow 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  visibility: visible;
  margin-left: 4px;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c3 > svg {
  display: block;
}

.c4 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css="
      width: 200px;
    "
    >
      <div
        className="c0"
      >
        <div
          className="c1"
        >
          Long Long Long Long Long Long Long Long Long Long Text
        </div>
        <div
          className="c2"
        >
          <div
            className="c3"
          >
            <div
              className="c4"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Sandbox/WithIcon Naive 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  min-width: 0;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  visibility: visible;
  margin-left: 4px;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c3 > svg {
  display: block;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 22px;
  height: 22px;
}

.c4::before {
  content: '';
  display: inline-block;
  height: 16px;
  width: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css={
        Array [
          "
          background-color: ",
          [Function],
          ";
        ",
        ]
      }
    >
      <div
        className="c0"
      >
        <div
          className="c1"
        >
          Menu
        </div>
        <div
          className="c2"
        >
          <div
            className="c3"
          >
            <div
              className="c4"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Sandbox/WithIcon Performance 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  min-width: 0;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 0;
  width: 16px;
  visibility: visible;
  margin-left: 4px;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: absolute;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c3 > svg {
  display: block;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 22px;
  height: 22px;
}

.c4::before {
  content: '';
  display: inline-block;
  height: 16px;
  width: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css={
        Array [
          "
          background-color: ",
          [Function],
          ";
        ",
        ]
      }
    >
      <div
        className="c0"
      >
        <div
          className="c1"
        >
          Menu
        </div>
        <div
          className="c2"
          width={16}
        >
          <div
            className="c3"
          >
            <div
              className="c4"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Sandbox/WithIcon Prefix 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  min-width: 0;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 0;
  width: 0px;
  visibility: visible;
  margin-right: 4px;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: absolute;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c2 > svg {
  display: block;
}

.c3 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background-color: currentColor;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
        ",
        [Function],
        "
        display: flex;
      ",
      ]
    }
  >
    <div
      css={
        Array [
          "
          background-color: ",
          [Function],
          ";
        ",
        ]
      }
    >
      <div
        className="c0"
      >
        <div
          className="c1"
          width={0}
        >
          <div
            className="c2"
          >
            <div
              className="c3"
            />
          </div>
        </div>
        <div
          className="c4"
        >
          Selection
        </div>
      </div>
    </div>
  </div>
</div>
`;
